runtime/lua/vim/_defaults.lua

```diff
@@ -98,18 +98,19 @@ do
   --- Map |gx| to call |vim.ui.open| on the <cfile> at cursor.
   do
     local function do_open(uri)
-      local cmd, err = vim.ui.open(uri)
-      local rv = cmd and cmd:wait(1000) or nil
-      if cmd and rv and rv.code ~= 0 then
-        err = ('vim.ui.open: command %s (%d): %s'):format(
+      local ok, cmd_or_err = vim.ui.open(uri)
+      local rv = ok and (cmd_or_err --[[@as vim.SystemObj]]):wait(1000) or nil
```

Comment:
typescript type system can model this so that `ok=true` would be understood to always have a valid `vim.SystemObj`, then this `--[[@as vim.SystemObj]]` wouldn't be needed. I wonder if LuaLS can do that